packages/compass-user-data/src/user-data.ts

```diff
@@ -7,26 +8,35 @@ import type { z } from 'zod';
 const { log, mongoLogId } = createLoggerAndTelemetry('COMPASS-USER-STORAGE');
 
 type SerializeContent<I> = (content: I) => string;
-type DeserializeContent = (content: string) => unknown;
+type DeserializeContent<I> = (content: string) => I;
 type GetFileName = (id: string) => string;
 
 export type UserDataOptions<Input> = {
   subdir: string;
   basePath?: string;
   serialize?: SerializeContent<Input>;
-  deserialize?: DeserializeContent;
+  deserialize?: DeserializeContent<Input>;
   getFileName?: GetFileName;
 };
 
-type ReadOptions = {
+type ReadStatsOptions<T> =
+  | {
+      readFileStats?: false;
+    }
+  | {
+      readFileStats: true;
+      mergeStats: (input: T, stats: Stats) => T;
+    };
+
+type ReadOptions<T> = ReadStatsOptions<T> & {
   ignoreErrors: boolean;
 };
 
 export class UserData<T extends z.Schema> {
   private readonly subdir: string;
   private readonly basePath?: string;
-  private readonly serialize: SerializeContent<T>;
-  private readonly deserialize: DeserializeContent;
+  private readonly serialize: SerializeContent<z.input<T>>;
+  private readonly deserialize: DeserializeContent<z.input<T>>;
```

Comment:
That's not true, at the moment this is called, nothing guarantees that the value returned is matching T